cmd/rebasehelpers/commitchecker/validate.go

```diff
@@ -0,0 +1,151 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+	"text/template"
+
+	"github.com/openshift/origin/cmd/rebasehelpers/util"
+)
+
+var UpstreamSummaryPattern = regexp.MustCompile(`UPSTREAM: ([\w-]+/[\w-]+: )?([0-9]{4,}:|<carry>:|<drop>:)`)
+var UpstreamRevertSummaryPattern = regexp.MustCompile(`UPSTREAM: revert: [a-f0-9]{7,}: ([\w-]+/[\w-]+: )?([0-9]{4,}:|<carry>:|<drop>:)`)
+
+var CommitSummaryErrorTemplate = `
+The following UPSTREAM commits have invalid summaries:
+
+{{ range .Commits }}  [{{ .Sha }}] {{ .Summary }}
+{{ end }}
+UPSTREAM commit summaries should look like:
+
+  UPSTREAM: [non-kube-account/repo: ]<PR number>|'<carry>'|'<drop>': description
+
+UPSTREAM commits which revert previous UPSTREAM commits should look like:
+
+  UPSTREAM: revert: <sha>: <normal upstream format>
+
+UPSTREAM commits are validated against the following regular expressions:
+
+  {{ .UpstreamSummaryPattern }}
+  {{ .UpstreamRevertSummaryPattern }}
+
+Examples of valid summaries:
+
+  UPSTREAM: 12345: An origin change
+  UPSTREAM: docker/docker: 12345: A docker change
+  UPSTREAM: <carry>: A carried kube change
+  UPSTREAM: <drop>: A dropped kube change
+  UPSTREAM: revert: abcd123: docker/docker: 12345: A docker change
+
+`
+
+// ValidateUpstreamCommitsWithoutGodepsChanges returns an error if any
+// upstream commits have no Godeps changes.
+func ValidateUpstreamCommitsWithoutGodepsChanges(commits []util.Commit) error {
+	problemCommits := []util.Commit{}
+	for _, commit := range commits {
+		if commit.HasGodepsChanges() && !commit.DeclaresUpstreamChange() {
+			problemCommits = append(problemCommits, commit)
+		}
+	}
+	if len(problemCommits) > 0 {
+		label := "The following commits contain Godeps changes but aren't declared as UPSTREAM:"
+		msg := renderGodepFilesError(label, problemCommits)
+		return fmt.Errorf(msg)
+	}
+	return nil
+}
+
+// ValidateUpstreamCommitModifiesSingleGodepsRepo returns an error if any
+// upstream commits have changes that span more than one Godeps repo.
+func ValidateUpstreamCommitModifiesSingleGodepsRepo(commits []util.Commit) error {
+	problemCommits := []util.Commit{}
+	for _, commit := range commits {
+		if !commit.HasGodepsChanges() {
+			continue
+		}
+		repos := map[string]struct{}{}
+		for _, file := range commit.Files {
+			if !file.HasGodepsChanges() {
+				continue
```

Comment:
An upstream commit should never touch a non-godep file, so this should be an error. I don't mind a separate validator.